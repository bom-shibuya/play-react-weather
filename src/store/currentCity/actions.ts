export const SWITCH_CITY = 'SWITCH_CITY'

export const switchCity = (payload: { city: string }) => ({
  payload,
  type: SWITCH_CITY as typeof SWITCH_CITY
})

export type Actions = ReturnType<typeof switchCity>
