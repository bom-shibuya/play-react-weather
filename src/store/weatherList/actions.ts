import { WeatherList } from './types'

export const FETCH_REQUEST = 'FETCH_REQUEST'
export const FETCH_SUCCESS = 'FETCH_SUCCESS'
export const FETCH_FAILURE = 'FETCH_FAILURE'

export const fetchRequest = (payload: { city: string }) => ({
  payload,
  type: FETCH_REQUEST as typeof FETCH_REQUEST
})

export interface IFetchSuccessPayload {
  city: string
  data: WeatherList
}

export const fetchSuccess = (payload: IFetchSuccessPayload) => ({
  payload,
  type: FETCH_SUCCESS as typeof FETCH_SUCCESS
})

export const fetchFailure = (payload: { city: string }) => ({
  payload,
  type: FETCH_FAILURE as typeof FETCH_FAILURE
})

export type Actions =
  | ReturnType<typeof fetchRequest>
  | ReturnType<typeof fetchSuccess>
  | ReturnType<typeof fetchFailure>
