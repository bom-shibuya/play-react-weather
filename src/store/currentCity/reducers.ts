import { Actions, SWITCH_CITY } from './actions'

export interface ICurrentCityState {
  currentCity: string
}

export const initialState: ICurrentCityState = {
  currentCity: 'Tokyo'
}

export const currentCity = (state = initialState, action: Actions) => {
  switch (action.type) {
    case SWITCH_CITY:
      return { ...state, currentCity: action.payload.city }
    default:
      return state
  }
}
