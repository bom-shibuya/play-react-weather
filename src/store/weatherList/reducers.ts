import { WeatherList } from '../../types'
import { Actions, FETCH_FAILURE, FETCH_REQUEST, FETCH_SUCCESS } from './actions'

export const initialState = {
  isError: false,
  isFetching: false,
  lastUpdateAt: null as null | number,
  list: [] as WeatherList
}

export type WeathersState = typeof initialState

export const weathers = (
  state = initialState,
  action: Actions
): WeathersState => {
  switch (action.type) {
    case FETCH_REQUEST:
      return { ...state, isFetching: true }
    case FETCH_SUCCESS:
      return {
        isError: false,
        isFetching: false,
        lastUpdateAt: Date.now(),
        list: [...action.payload.data]
      }
    case FETCH_FAILURE:
      return {
        ...state,
        isError: true,
        lastUpdateAt: Date.now()
      }
    default:
      return state
  }
}

export interface IWetherListState {
  [key: string]: WeathersState
}

export const weatherList = (state = {}, action: Actions): IWetherListState => {
  switch (action.type) {
    case FETCH_REQUEST:
    case FETCH_SUCCESS:
    case FETCH_FAILURE:
      return {
        ...state,
        [action.payload.city]: weathers(initialState, action)
      }
    default:
      return state
  }
}
