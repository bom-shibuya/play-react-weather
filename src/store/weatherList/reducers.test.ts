import snapshotDiff from 'snapshot-diff'
import { fetchFailure, fetchRequest, fetchSuccess } from './actions'
import { initialState, weatherList, weathers, WeathersState } from './reducers'

describe('reducer weathers test', () => {
  beforeEach(() => {
    Date.now = jest.fn(() => new Date(Date.UTC(2017, 7, 9, 8)).valueOf())
  })

  const payload = {
    city: 'Tokyo'
  }

  test('shold return initial state', () => {
    expect(
      snapshotDiff(initialState, weathers(initialState, {} as any))
    ).toMatchSnapshot()
  })

  test('shold return state tokyo isFetching true', () => {
    expect(
      snapshotDiff(initialState, weathers(initialState, fetchRequest(payload)))
    ).toMatchSnapshot()
  })

  test('shold return state tokyo include data', () => {
    expect(
      snapshotDiff(
        initialState,
        weathers(
          initialState,
          fetchSuccess({
            ...payload,
            data: [{ test: 'expected value' } as any]
          })
        )
      )
    ).toMatchSnapshot()
  })

  test('shold return state tokyo isError true', () => {
    expect(
      snapshotDiff(initialState, weathers(initialState, fetchFailure(payload)))
    ).toMatchSnapshot()
  })
})

describe('reducer weatherList test', () => {
  beforeEach(() => {
    Date.now = jest.fn(() => new Date(Date.UTC(2017, 7, 9, 8)).valueOf())
  })

  const payload = {
    city: 'Tokyo'
  }

  test('should return initial state', () => {
    expect(snapshotDiff({}, weatherList({}, {} as any))).toMatchSnapshot()
  })

  test('shold return state tokyo isFetching true', () => {
    expect(
      snapshotDiff({}, weatherList({}, fetchRequest(payload)))
    ).toMatchSnapshot()
  })

  test('shold return state tokyo include data', () => {
    expect(
      snapshotDiff(
        {},
        weatherList(
          {},
          fetchSuccess({
            ...payload,
            data: [{ test: 'expected value' } as any]
          })
        )
      )
    ).toMatchSnapshot()
  })

  test('shold return state tokyo isError true', () => {
    expect(
      snapshotDiff({}, weatherList({}, fetchFailure(payload)))
    ).toMatchSnapshot()
  })
})
