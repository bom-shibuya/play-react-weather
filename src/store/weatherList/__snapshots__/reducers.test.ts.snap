// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer weatherList test shold return state tokyo include data 1`] = `
"Snapshot Diff:
- First value
+ Second value

- Object {}
+ Object {
+   \\"Tokyo\\": Object {
+     \\"isError\\": false,
+     \\"isFetching\\": false,
+     \\"lastUpdateAt\\": 1502265600000,
+     \\"list\\": Array [
+       Object {
+         \\"test\\": \\"expected value\\",
+       },
+     ],
+   },
+ }"
`;

exports[`reducer weatherList test shold return state tokyo isError true 1`] = `
"Snapshot Diff:
- First value
+ Second value

- Object {}
+ Object {
+   \\"Tokyo\\": Object {
+     \\"isError\\": true,
+     \\"isFetching\\": false,
+     \\"lastUpdateAt\\": null,
+     \\"list\\": Array [],
+   },
+ }"
`;

exports[`reducer weatherList test shold return state tokyo isFetching true 1`] = `
"Snapshot Diff:
- First value
+ Second value

- Object {}
+ Object {
+   \\"Tokyo\\": Object {
+     \\"isError\\": false,
+     \\"isFetching\\": true,
+     \\"lastUpdateAt\\": null,
+     \\"list\\": Array [],
+   },
+ }"
`;

exports[`reducer weatherList test should return initial state 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`reducer weathers test shold return initial state 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`reducer weathers test shold return state tokyo include data 1`] = `
"Snapshot Diff:
- First value
+ Second value

  Object {
    \\"isError\\": false,
    \\"isFetching\\": false,
-   \\"lastUpdateAt\\": null,
-   \\"list\\": Array [],
+   \\"lastUpdateAt\\": 1502265600000,
+   \\"list\\": Array [
+     Object {
+       \\"test\\": \\"expected value\\",
+     },
+   ],
  }"
`;

exports[`reducer weathers test shold return state tokyo isError true 1`] = `
"Snapshot Diff:
- First value
+ Second value

  Object {
-   \\"isError\\": false,
+   \\"isError\\": true,
    \\"isFetching\\": false,
    \\"lastUpdateAt\\": null,
    \\"list\\": Array [],
  }"
`;

exports[`reducer weathers test shold return state tokyo isFetching true 1`] = `
"Snapshot Diff:
- First value
+ Second value

  Object {
    \\"isError\\": false,
-   \\"isFetching\\": false,
+   \\"isFetching\\": true,
    \\"lastUpdateAt\\": null,
    \\"list\\": Array [],
  }"
`;
